### PR TITLE
track progress in IndexFeatureFile

### LINKED ISSUE #1307: IndexFeatureFile prints no progress
IndexFeatureFile prints no progress and so it seems like it's hanging for a long time

### PR DESCRIPTION
**Pull request by akiezun**
fixes #1307 

@droazen can you review?

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/IndexFeatureFile.java [hunk 1]
```diff
@@ -84,6 +84,78 @@ protected Object doWork() {
         return indexFile.getAbsolutePath();
     }
 
+    /**
+     * We want to report progress when indexing but the indexing code is in htsjdk and not available to us directly.
+     * The workaround is to make a special decorator 'codec' that gets called on every decoded Feature and this can use used to track progress.
+     * The codec delegates all calls and reports progress as it goes.
+     */
+    private static final class ProgressReportingDelegatingCodec<A extends Feature, B> implements FeatureCodec<A, B>{
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/IndexFeatureFile.java]: Don't embed this class within the indexing utility -- make public and move to `utils/codecs`

File: src/main/java/org/broadinstitute/hellbender/tools/IndexFeatureFile.java [hunk 2]
```diff
@@ -84,6 +84,78 @@ protected Object doWork() {
...
+        private final FeatureCodec<A, B> delegatee;
+
+        private final ProgressMeter pm;
+
+        ProgressReportingDelegatingCodec(final FeatureCodec<A, B> delegatee){
+            this.delegatee = delegatee;
+            this.pm = new ProgressMeter();
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/IndexFeatureFile.java]: Allow the ProgressMeter's "seconds between updates" setting to be configurable via the `ProgressReportingDelegatingCodec` constructor, and make constructor public.

File: src/main/java/org/broadinstitute/hellbender/tools/IndexFeatureFile.java [hunk 3]
```diff
@@ -84,6 +84,78 @@ protected Object doWork() {
...
+        @Override
+        public Feature decodeLoc(final B b) throws IOException {
+            final Feature f = delegatee.decodeLoc(b);
+            pm.update(f);
+            return f;
+        }
+
+        @Override
+        public A decode(final B b) throws IOException {
+            return delegatee.decode(b);
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/IndexFeatureFile.java]: Update the `ProgressMeter` when `decode()` is called as well

File: src/main/java/org/broadinstitute/hellbender/tools/IndexFeatureFile.java [hunk 4]
```diff
@@ -84,6 +84,78 @@ protected Object doWork() {
...
+
+        @Override
+        public Class<A> getFeatureType() {
+            return delegatee.getFeatureType();
+        }
+
+        @Override
+        public B makeSourceFromStream(final InputStream bufferedInputStream) {
+            //This is the first call to the codes so we start the progress monitor here.
+            pm.start();
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/IndexFeatureFile.java]: This doesn't seem 100% reliable in the general case -- recommend starting the progress meter in `decode()/`decodeLoc()` instead, and keep track of whether the meter has already been started.

File: src/main/java/org/broadinstitute/hellbender/tools/IndexFeatureFile.java [hunk 5]
```diff
@@ -127,6 +199,15 @@ else if ( featureFile.getName().endsWith(GVCF_FILE_EXTENSION) ) {
     }
 
     private boolean isVCFCodec( final FeatureCodec<? extends Feature, ?> codec ) {
-        return codec.getClass() == VCFCodec.class || codec.getClass() == VCF3Codec.class;
+        if (codec.getClass() == VCFCodec.class){
+            return true;
+        }
+        if (codec.getClass() == VCF3Codec.class){
+            return true;
+        }
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/IndexFeatureFile.java]: if ( codec.getClass() == VCFCodec.class || codec.getClass() == VCF3Codec.class ) {
        return true;
    }

**Comment by droazen** [comment]
Review complete -- back to @akiezun

File: src/main/java/org/broadinstitute/hellbender/tools/IndexFeatureFile.java (hunk 1 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/IndexFeatureFile.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/IndexFeatureFile.java (hunk 2 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/IndexFeatureFile.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/IndexFeatureFile.java (hunk 3 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/IndexFeatureFile.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/IndexFeatureFile.java (hunk 5 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/IndexFeatureFile.java]: ok

**Comment by akiezun** [comment]
done - back to @droazen

**Comment by akiezun** [comment]
@lbergelson says the cloud failure is expected because of the keys. Don't let this stop the review, @droazen

File: src/main/java/org/broadinstitute/hellbender/engine/ProgressMeter.java [hunk 6]
```diff
@@ -153,8 +176,15 @@ public void start() {
      * statistics to the logger roughly every {@link #secondsBetweenUpdates} seconds.
      *
      * @param currentLocus the genomic location of the record just processed or null if the most recent record had no location.
+     * @throws IllegalStateException is the meter has not been started yet
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/ProgressMeter.java]: is -> if

File: src/main/java/org/broadinstitute/hellbender/engine/ProgressMeter.java [hunk 7]
```diff
@@ -169,8 +199,16 @@ public void update( final Locatable currentLocus ) {
 
     /**
      * Stop the progress meter and output summary statistics to the logger
+     * @throws IllegalStateException is the meter has not been started yet
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/ProgressMeter.java]: is -> if

File: src/main/java/org/broadinstitute/hellbender/engine/ProgressMeter.java (hunk 6 above)
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/ProgressMeter.java]: "or has been stopped"

File: src/main/java/org/broadinstitute/hellbender/engine/ProgressMeter.java (hunk 7 above)
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/ProgressMeter.java]: "or has been stopped"

File: src/main/java/org/broadinstitute/hellbender/utils/codecs/ProgressReportingDelegatingCodec.java [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+            throw new IllegalArgumentException("secondsBetweenUpdates must be > 0.0");
+        }
+        this.delegatee = delegatee;
+        this.pm = new ProgressMeter(secondsBetweenUpdates);
+    }
+
+    @Override
+    public Feature decodeLoc(final B b) throws IOException {
+        if (delegatee == null) {
+            throw new IllegalStateException("this progress monitor cannot be used without a delegatee.");
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/codecs/ProgressReportingDelegatingCodec.java]: replace all instances of "this progress monitor" in the error messages in this class with "this codec"

File: src/main/java/org/broadinstitute/hellbender/utils/codecs/ProgressReportingDelegatingCodec.java [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+    public void close(final B b) {
+        if (delegatee == null) {
+            throw new IllegalStateException("this progress monitor cannot be used without a delegatee.");
+        }
+        delegatee.close(b);
+    }
+
+    @Override
+    public boolean canDecode(final String path) {
+        //If there's not delegatee then we're going to say no to all questions here
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/codecs/ProgressReportingDelegatingCodec.java]: not -> no

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/bqsr/BaseRecalibratorIntegrationTest.java [hunk 10]
```diff
@@ -14,7 +14,7 @@
 import java.io.IOException;
 import java.util.Arrays;
 
-public final class BaseRecalibratorIntegrationTest extends CommandLineProgramTest{
+public final class  BaseRecalibratorIntegrationTest extends CommandLineProgramTest{
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/bqsr/BaseRecalibratorIntegrationTest.java]: revert whitespace change here

**Comment by droazen** [comment]
:+1: minor remaining comments only, merge after addressing. back to @akiezun

File: src/main/java/org/broadinstitute/hellbender/engine/ProgressMeter.java (hunk 6 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/engine/ProgressMeter.java]: fixed both

File: src/main/java/org/broadinstitute/hellbender/engine/ProgressMeter.java (hunk 7 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/engine/ProgressMeter.java]: fixed both

File: src/main/java/org/broadinstitute/hellbender/utils/codecs/ProgressReportingDelegatingCodec.java (hunk 8 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/codecs/ProgressReportingDelegatingCodec.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/codecs/ProgressReportingDelegatingCodec.java (hunk 9 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/codecs/ProgressReportingDelegatingCodec.java]: fixed

**Pull request merged**